File: leetcode/C++/Reorder_list.cpp
#include<iostream>
#include<string>
#include<vector>

using namespace std;

struct  ListNode{
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL){}
};

class Solution {
	public:
		void reorderList(ListNode *head) {
			ListNode *cur = head;
			vector<int> res;

			if(!cur)
				return;

			while(cur)
			{
				res.push_back(cur->val);
				cur = cur->next;
			}

			for(int i = 0; i <=  (res.size()+1)/2; i++)
			{
				if(i==0)
				{
					head->val = res[i];
					cur = head;
					if(i < (res.size()-i-1))
					{
						cur->next = new ListNode(res[res.size()-1]);
						cur = cur->next;
					}
				}
				else
				{
					if(i < (res.size()-i-1))
					{
						cur->next = new ListNode(res[i]);
						cur = cur->next;
						cur->next = new ListNode(res[res.size()-i]);
						cur = cur->next;
					}
					if(i == (res.size()-i-1))
					{
						cur ->next = new ListNode(res[i]);
						cur = cur->next;
					}
				}
			}

			cur->next = NULL;
		}
};

int main(void)
{
	Solution s;
	
	ListNode *head;
	head = new ListNode(1);
	s.reorderList(head);

	cout << head->val  << head->next->val << head->next->next->val << endl; 

	return 0;
}

File: leetcode/linklist/reorder list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func reorderList(head *ListNode)  {
    count := 0
    cur := head
    for;cur != nil;{
        cur = cur.Next 
        count++
    }

    firstTail := findKth(head, (count+1)/2)
    secondHead := firstTail.Next
    firstTail.Next = nil 
    secondHead = reverse(secondHead)

    cur1, cur2 := head, secondHead
    for;cur1 != nil && cur2 != nil;{
        tmp1 := cur1.Next
        tmp2 := cur2.Next
        cur1.Next = cur2 
        cur2.Next = tmp1 
        cur1 = tmp1 
        cur2 = tmp2
    }
}

func reverse(head *ListNode) *ListNode{
    var pre *ListNode
    cur := head 
    
    for;cur != nil;{
        tmp := cur.Next 
        cur.Next = pre 
        pre = cur
        cur = tmp
    }
    return pre
}

func findKth(head *ListNode, k int) *ListNode{
    cur := head
    for;cur != nil;{
        k--
        if k == 0{
            return cur
        }
        cur = cur.Next
    }
    return nil
}

